os/net/mac/csma/csma-security.c

```diff
@@ -0,0 +1,267 @@
+    LOG_INFO_(" %u (%u) KEY:0x%02x\n", packetbuf_datalen(), packetbuf_totlen(),
+              packetbuf_attr(PACKETBUF_ATTR_KEY_INDEX));
+  }
+  return hdr_len;
+}
+
+/*---------------------------------------------------------------------------*/
+int
+csma_security_frame_len(void)
+{
+  if(packetbuf_attr(PACKETBUF_ATTR_SECURITY_LEVEL) > 0 &&
+     packetbuf_attr(PACKETBUF_ATTR_KEY_INDEX) != 0xffff) {
+    return NETSTACK_FRAMER.length() + MIC_LEN;
+  }
+  return NETSTACK_FRAMER.length();
+}
+/*---------------------------------------------------------------------------*/
+int
+csma_security_parse_frame(void)
+{
+  int hdr_len;
+
+  hdr_len = NETSTACK_FRAMER.parse();
+  if(hdr_len < 0) {
+    return hdr_len;
+  }
+
+  if(packetbuf_attr(PACKETBUF_ATTR_SECURITY_LEVEL) == 0) {
+    /* No security - no more processing required */
+    return hdr_len;
+  }
+
+  LOG_INFO("LLSEC-IN: ");
+  LOG_INFO_LLADDR(packetbuf_addr(PACKETBUF_ADDR_SENDER));
+  LOG_INFO_(" ");
+  LOG_INFO_LLADDR(packetbuf_addr(PACKETBUF_ADDR_RECEIVER));
+  LOG_INFO_(" %d %u (%u) LV:%d KM:%d KEY:0x%02x\n", hdr_len, packetbuf_datalen(),
+            packetbuf_totlen(), packetbuf_attr(PACKETBUF_ATTR_SECURITY_LEVEL),
+            packetbuf_attr(PACKETBUF_ATTR_KEY_ID_MODE),
+            packetbuf_attr(PACKETBUF_ATTR_KEY_INDEX));
+
+  if(packetbuf_attr(PACKETBUF_ATTR_SECURITY_LEVEL) != CSMA_LLSEC_SECURITY_LEVEL) {
+    LOG_INFO("received frame with wrong security level (%u) from ",
+           packetbuf_attr(PACKETBUF_ATTR_SECURITY_LEVEL));
+    LOG_INFO_LLADDR(packetbuf_addr(PACKETBUF_ADDR_SENDER));
+    LOG_INFO_("\n");
+    return FRAMER_FAILED;
+  }
+
+  if(packetbuf_attr(PACKETBUF_ATTR_KEY_ID_MODE) != CSMA_LLSEC_KEY_ID_MODE) {
+    LOG_INFO("received frame with wrong key id mode (%u) from ",
+           packetbuf_attr(PACKETBUF_ATTR_KEY_ID_MODE));
+    LOG_INFO_LLADDR(packetbuf_addr(PACKETBUF_ADDR_SENDER));
+    LOG_INFO("\n");
+    return FRAMER_FAILED;
+  }
+
+  if(linkaddr_cmp(packetbuf_addr(PACKETBUF_ADDR_SENDER), &linkaddr_node_addr)) {
+    LOG_INFO("frame from ourselves\n");
+    return FRAMER_FAILED;
+  }
+
+  if(packetbuf_datalen() <= MIC_LEN) {
+    LOG_ERR("MIC error - too little data in frame!\n");
+    return FRAMER_FAILED;
+  }
+
+  packetbuf_set_datalen(packetbuf_datalen() - MIC_LEN);
+  if(!aead(hdr_len, 0)) {
+    LOG_INFO("received unauthentic frame %u from ",
+             (unsigned int) anti_replay_get_counter());
+    LOG_INFO_LLADDR(packetbuf_addr(PACKETBUF_ADDR_SENDER));
+    LOG_INFO_("\n");
+    return FRAMER_FAILED;
+  }
+
+  /* TODO anti-reply protection */
+  return hdr_len;
+}
+/*---------------------------------------------------------------------------*/
+#else
+/* The "unsecure" version of the create frame / parse frame */
+int
+csma_security_create_frame(void)
+{
+  packetbuf_set_attr(PACKETBUF_ATTR_FRAME_TYPE, FRAME802154_DATAFRAME);
+  return NETSTACK_FRAMER.create();
+}
+int
+csma_security_parse_frame(void)
+{
+  return NETSTACK_FRAMER.parse();
+}
```

Comment:
Should the else block include a empty version of `csma_security_set_key/2`? It's exported in `csma.h` and is expected to be available.